File: backend.scala/app/com/skewwhiffy/auraltester/notes/Octave.scala
package com.skewwhiffy.auraltester.notes

object Octave {
  lazy val default: Octave = Octave(0)
}

case class Octave(offsetFromDefault: Int) {
  lazy val up: Octave = Octave(offsetFromDefault + 1)

  lazy val down: Octave = Octave(offsetFromDefault - 1)

  def >(other: Octave): Boolean = offsetFromDefault > other.offsetFromDefault

  def <(other: Octave): Boolean = offsetFromDefault < other.offsetFromDefault

  def >=(other: Octave): Boolean = offsetFromDefault > other.offsetFromDefault || this == other

  def <=(other: Octave): Boolean = offsetFromDefault < other.offsetFromDefault || this == other
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/controller/Octave.kt
package com.skewwhiffy.auraltester.notation.model.note

// @JvmRecord
data class Octave(val offsetFromDefault: Int) : Comparable<Octave?> {
    companion object {
        val default: Octave
            get() = Octave(0)
    }
    /*
    fun toDao(): OctaveDao {
        return OctaveDao(offsetFromDefault)
    }

    fun up(): Octave {
        return Octave(offsetFromDefault + 1)
    }

    fun down(): Octave {
        return Octave(offsetFromDefault - 1)
    }
    */

    override fun compareTo(other: Octave?): Int {
        if (other == null) {
            return 1
        }
        return offsetFromDefault - other.offsetFromDefault
    }

}
